keygen.js

```diff
@@ -0,0 +1,314 @@
+	function wipe(arr) {
+		for (var i = 0; i < arr.length; i++)
+			arr[i] = 0xFE;
+	}
+
+	function nextPowerOfTwo(v) {
+		var p = 2;
+		while ((v >>= 1)) {
+			p <<= 1;
+		}
+		return p;
+	}
+
+	// openssl.js suggests using wasmfs, however, it adds a
+	// lot of dependencies and is difficult to package, so
+	// just implement a basic virtual FS here.
+	let fs = { }; // TODO: should have temp scope
+	let lineBuf = '';
+	let tty = {
+		write : function(arr) {
+			for (var c of arr)
+				if (c == 10) {
+					console.log(lineBuf); lineBuf = '';
+				} else {
+					lineBuf += String.fromCharCode(c);
+				}
+		}
+	};
+	let fds = [ { }, tty, tty ];
+	let fsImpl = {
+		existsSync : function(fn) {
+			// console.log('existsSync', fn);
+			return fn in fs;
+		},
+		realpathSync : function(n) {
+			return n;
+		},
+		mkdirSync : function() {
+			return false;
+		},
+		openSync : function(fn, flags) {
+			// console.log('openSync', fn, flags);
+			var ro = flags & 1;
+			if (!ro)
+				fs[fn] = new Uint8Array(0);
+			else
+				if (!(fn in fs))
+					return -1;
+			var pos = 0;
+			fds.push({
+				read : function(length) {
+					var arr = fs[fn].subarray(pos, pos + length);
+					pos += arr.length;
+					return arr;
+				},
+				write : ro ? null : function(arr) {
+					// console.log('write', arr.length, '@', pos, '/', fs[fn].buffer.byteLength);
+					if (fs[fn].buffer.byteLength < pos + arr.length) {
+						var newFile = new Uint8Array(nextPowerOfTwo(pos + arr.length));
+						newFile.set(fs[fn]);
+						wipe(new Uint8Array(fs[fn].buffer));
+						fs[fn] = newFile.subarray(0, pos);
+					}
+					var target = new Uint8Array(fs[fn].buffer, pos, arr.length);
+					target.set(arr);
+					pos += arr.length;
+					fs[fn] = new Uint8Array(fs[fn].buffer, 0, pos);
+					wipe(arr);
+				},
+			});
+			return fds.length - 1;
+		},
+		readSync : function(fd, buffer, offset, length, position) {
+			// console.log(fd, buffer, offset, length);
+			if (fd < 0) return 0;
+			var arr = fds[fd].read(length);
+			buffer.subarray(offset, offset + arr.length).set(arr);
+			return arr.length;
+		},
+		writeSync : function(fd, buffer) {
+			if (fd < 0) return 0;
+			fds[fd].write(buffer);
+			return buffer.length;
+		},
+		closeSync : function(fd) {
+			if (fd < 0) return;
+			fds[fd] = null;
+		},
+		fstatSync : function() {
+			return {
+				isBlockDevice : () => false,
+				isCharacterDevice : () => false,
+				isDirectory : () => false,
+				isFIFO : () => false,
+				isFile : () => true,
+			};
+		},
+		constants : {
+			O_RDONLY : 1,
+		},
+	};
+
+	let openSSL = new OpenSSL({
+		fs: fsImpl,
+		rootDir: '/'
+	});
+
+	async function genSpkac(algorithm, pkeyopts) {
+		if (algorithm === undefined)
+			algorithm = 'RSA';
+		if (!algorithm.length || algorithm.match(/[\s]/))
+			throw 'Invalid algorithm';
+		pkeyopts = pkeyopts.split(/[\s]{1,}/g).filter(Boolean);
+
+		await openSSL.runCommand(
+			"genpkey" +
```

Comment:
RSA defaults to 1024 bits, this should specify 4096 bits

Reply:
MDN says:

> For RSA keys, the keyparams parameter is not used (ignored if present). The user may be given a choice of RSA key strengths. Currently, the user is given a choice between "high" strength (2048 bits) and "medium" strength (1024 bits).

0. We should pick up `keytype` (and maybe other properties) from the tag
1. We can add a `keysize` parameter which translates to `rsa_keygen_bits:XXX`, or allow setting `keyopts` on the tag which would get directly concatenated to the value in the `keyopts` input field. (The input field can be hidden with CSS if the site doesn't want users to customize options.)

Reply:
Well, for my purposes I don't want people to choose and only want 4096 bits

Reply:
Yep, we can do that but as this library aims to be a generic polyfill it probably should stick as close to the behaviour of `<keygen>` as possible. We can customize for our needs on site.

Reply:
> We should pick up `keytype` (and maybe other properties) from the tag

Done

> We can add a `keysize` parameter which translates to `rsa_keygen_bits:XXX`, or allow setting `keyopts` on the tag which would get directly concatenated to the value in the `keyopts` input field.

Added `-keygenjs-pkeyopts` with the same semantics as the value of the options input (and `-pkeyopts` OpenSSL switch).